src/main/java/htsjdk/samtools/cram/structure/block/Block.java

```diff
@@ -0,0 +1,272 @@
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ****************************************************************************
+ */
+package htsjdk.samtools.cram.structure.block;
+
+import htsjdk.samtools.cram.CRAMException;
+import htsjdk.samtools.cram.common.CramVersions;
+import htsjdk.samtools.cram.compression.ExternalCompression;
+import htsjdk.samtools.cram.compression.ExternalCompressor;
+import htsjdk.samtools.cram.io.*;
+import htsjdk.samtools.util.RuntimeIOException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+/**
+ * Class representing CRAM block concept and some methods to operate with block content. CRAM block is used to hold some (usually
+ * homogeneous) binary data. An external compression can be applied to the content of a block. The class provides some instantiation static
+ * methods, for example to read a block from an input stream. Blocks can be written out to an output stream, this may be considered as a way
+ * to serialize/deserialize blocks.
+ */
+public class Block {
+    /**
+     * Only external blocks have meaningful Content IDs
+     * Other blocks are required to have a Content ID of 0
+     */
+    public static final int NO_CONTENT_ID = 0;
+
+    /**
+     * Compression method that applied to this block's content.
+     */
+    private final BlockCompressionMethod method;
+
+    /**
+     * Identifies CRAM content type of the block.
+     */
+    private final BlockContentType contentType;
+
+    /**
+     * The content stored in this block, in compressed form (if applicable)
+     */
+    protected final byte[] compressedContent;
+
+    /**
+     * The length of the content stored in this block when uncompressed
+     */
+    private final int uncompressedLength;
+
+    /**
+     * Private constructor of a generic Block, to be called by static factory methods and subclasses.
+     *
+     * @param method the block compression method.  Can be RAW, if uncompressed
+     * @param type whether this is a header or data block, and which kind
+     * @param compressedContent the compressed form of the data to be stored in this block
+     * @param uncompressedLength the length of the content stored in this block when uncompressed
+     */
+    protected Block(final BlockCompressionMethod method,
+                    final BlockContentType type,
+                    final byte[] compressedContent,
+                    final int uncompressedLength) {
+        this.method = method;
+        this.contentType = type;
+        this.compressedContent = compressedContent;
+        this.uncompressedLength = uncompressedLength;
```

Comment:
It seems error prone to mix the uncompressed content and compressed lengths like this, especially when many of the calls to this pass in raw content as the compressed content.

Maybe this should have two separate constructors:

`Block (final BlockContentType type, final byte[] rawContent)` (for creating raw blocks from raw content)

and:

`Block (final BlockContentType type, final BlockCompressionMethod method, final byte[] compressedContent, final int expectedUncompressedLength)`

and just have corresponding factory methods (the call sites can pass in the contentType). Then have the implementation keep the expectedUncompressedLength length separate from the actual length, and manage the content and length inter-conversions on demand. We'd need to do a similar thing for `External`.

Reply:
ok for Block.  Not sure this makes sense for External - which now has two constructors for different purposes already.